File: crates/pray-cli/src/confess.rs
use crate::{
    current_signer, lockfile_path, locked_package, manifest_path, resolve_project,
};
use pray_core::hashing::sha256_prefixed;
use pray_core::lockfile::read_lockfile;
use pray_core::registry::{submit_confession, ConfessionSubmission};
use pray_core::{PrayError, PrayResult};
use std::time::{SystemTime, UNIX_EPOCH};

pub(crate) fn confess_command(
    package: Option<String>,
    from_lock: Option<String>,
    version: Option<String>,
    accepted: bool,
    rejected: bool,
    note: Option<String>,
    url: Option<String>,
) -> PrayResult<()> {
    if accepted == rejected {
        return Err(PrayError::Unsupported(
            "confess requires exactly one of --accepted or --rejected".to_string(),
        ));
    }

    let project = resolve_project(&manifest_path())?;
    let lockfile = read_lockfile(&lockfile_path()).ok();

    let (package_name, resolved_version, package_resolution) = if let Some(span_id) = from_lock {
        let lockfile = lockfile.as_ref().ok_or_else(|| {
            PrayError::Resolution("confess --from-lock requires an existing lockfile".to_string())
        })?;
        let span = lockfile
            .managed_span
            .iter()
            .find(|record| record.id == span_id)
            .ok_or_else(|| PrayError::Resolution(format!("lockfile span {span_id} not found")))?;
        let package_resolution = project
            .packages
            .iter()
            .find(|resolved_package| resolved_package.declaration.name == span.package)
            .ok_or_else(|| PrayError::Resolution(format!("package {} not found", span.package)))?;
        let locked_package = locked_package(lockfile, package_resolution).ok_or_else(|| {
            PrayError::Resolution(format!("lockfile package {} not found", span.package))
        })?;
        let resolved_version = match version {
            Some(requested_version) if requested_version != locked_package.version => {
                return Err(PrayError::Resolution(format!(
                    "lockfile span {} version {} does not match requested version {}",
                    span_id, locked_package.version, requested_version
                )));
            }
            Some(requested_version) => requested_version,
            None => locked_package.version.clone(),
        };
        (span.package.clone(), resolved_version, package_resolution)
    } else {
        let package_name = package
            .ok_or_else(|| PrayError::Unsupported("confess requires a package name".to_string()))?;
        let package_resolution = project
            .packages
            .iter()
            .find(|resolved_package| resolved_package.declaration.name == package_name)
            .ok_or_else(|| PrayError::Resolution(format!("package {package_name} not found")))?;
        let resolved_version = match version {
            Some(requested_version) if requested_version != package_resolution.spec.version => {
                return Err(PrayError::Resolution(format!(
                    "package {package_name} version {} does not match resolved version {}",
                    requested_version, package_resolution.spec.version
                )));
            }
            Some(requested_version) => requested_version,
            None => package_resolution.spec.version.clone(),
        };
        (package_name, resolved_version, package_resolution)
    };

    let source_name = package_resolution
        .declaration
        .source
        .as_ref()
        .ok_or_else(|| {
            PrayError::Resolution(format!("package {package_name} is missing a source"))
        })?;
    let source_url = if let Some(url) = url {
        url
    } else {
        project
            .manifest
            .sources
            .iter()
            .find(|source| source.name == *source_name)
            .map(|source| source.url.clone())
            .ok_or_else(|| PrayError::Resolution(format!("unknown source: {source_name}")))?
    };

    let lockfile_reference = lockfile
        .as_ref()
        .and_then(|lockfile| lockfile.file_hash().ok());
    let timestamp = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|error| PrayError::Resolution(error.to_string()))?
        .as_secs()
        .to_string();
    let mut confession = ConfessionSubmission {
        package: package_name.clone(),
        version: resolved_version,
        status: if accepted {
            "accepted".to_string()
        } else {
            "rejected".to_string()
        },
        note,
        lockfile: lockfile_reference,
        distribution_point: Some(source_url.clone()),
        signer: Some(current_signer()?),
        timestamp: Some(timestamp),
        signature: None,
    };
    let signature_payload =
        serde_json::to_vec(&confession).map_err(|error| PrayError::Manifest(error.to_string()))?;
    confession.signature = Some(sha256_prefixed(&signature_payload));
    submit_confession(&source_url, &confession)?;
    println!(
        "Confession submitted for {} {}",
        confession.package, confession.version
    );
    Ok(())
}

File: rubygems/pray-cli/lib/pray/confess.rb
# frozen_string_literal: true

require "json"
require_relative "session"

module Pray
  ConfessionSubmission = Struct.new(
    :package, :version, :status, :note, :lockfile,
    :distribution_point, :signer, :timestamp, :signature,
    keyword_init: true
  )

  module Confess
    module_function

    def submit(package:, from_lock:, version:, accepted:, rejected:, note:, url:,
      project_root: Dir.pwd)
      raise Error.unsupported("confess requires exactly one of --accepted or --rejected") if accepted == rejected

      manifest_path = File.join(project_root, "Prayfile")
      project = Resolve.resolve_project(manifest_path)
      lockfile_path = File.join(project_root, "Prayfile.lock")
      lockfile = File.file?(lockfile_path) ? Lockfile.read_lockfile(lockfile_path) : nil

      package_name, resolved_version, package_resolution =
        resolve_target(project, lockfile, package, from_lock, version)
      source_url = resolve_source_url(project, package_resolution, package_name, url)
      if source_url.start_with?("pray+ssh://", "ssh+pray://")
        raise Error.unsupported("pray_ssh confession submit is not implemented yet in pray-cli Ruby")
      end

      confession = ConfessionSubmission.new(
        package: package_name,
        version: resolved_version,
        status: accepted ? "accepted" : "rejected",
        note: note,
        lockfile: lockfile&.file_hash,
        distribution_point: source_url,
        signer: Session.current_signer(project_root),
        timestamp: Time.now.to_i.to_s,
        signature: nil
      )
      confession.signature = Hashing.sha256_prefixed(JSON.generate(confession_to_hash(confession)))
      Registry.http_post(
        Registry.join_url(source_url, "v1/confessions"),
        "application/json",
        JSON.generate(confession_to_hash(confession))
      )
      puts "Confession submitted for #{confession.package} #{confession.version}"
    end

    def resolve_target(project, lockfile, package, from_lock, version)
      if from_lock
        raise Error.resolution("confess --from-lock requires an existing lockfile") unless lockfile

        span = lockfile.managed_span.find { |record| record.id == from_lock }
        raise Error.resolution("lockfile span #{from_lock} not found") unless span

        package_resolution = project.packages.find { |entry| entry.declaration.name == span.package }
        raise Error.resolution("package #{span.package} not found") unless package_resolution

        locked = lockfile.package.find { |entry| entry.name == span.package }
        raise Error.resolution("lockfile package #{span.package} not found") unless locked

        if version && version != locked.version
          raise Error.resolution(
            "lockfile span #{from_lock} version #{locked.version} does not match requested version #{version}"
          )
        end
        [span.package, version || locked.version, package_resolution]
      else
        raise Error.unsupported("confess requires a package name") unless package

        package_resolution = project.packages.find { |entry| entry.declaration.name == package }
        raise Error.resolution("package #{package} not found") unless package_resolution

        if version && version != package_resolution.spec.version
          raise Error.resolution(
            "package #{package} version #{version} does not match resolved version #{package_resolution.spec.version}"
          )
        end
        [package, version || package_resolution.spec.version, package_resolution]
      end
    end

    def resolve_source_url(project, package_resolution, package_name, url)
      return url if url

      source_name = package_resolution.declaration.source
      raise Error.resolution("package #{package_name} is missing a source") unless source_name

      source = project.manifest.sources.find { |entry| entry.name == source_name }
      raise Error.resolution("unknown source: #{source_name}") unless source

      source.url
    end

    def confession_to_hash(confession)
      hash = {
        "package" => confession.package,
        "version" => confession.version,
        "status" => confession.status
      }
      hash["note"] = confession.note if confession.note
      hash["lockfile"] = confession.lockfile if confession.lockfile
      hash["distribution_point"] = confession.distribution_point if confession.distribution_point
      hash["signer"] = confession.signer if confession.signer
      hash["timestamp"] = confession.timestamp if confession.timestamp
      hash["signature"] = confession.signature if confession.signature
      hash
    end
  end
end
